File: src/dotnet/src/Drawbridge.Core/Exec/Http.cs
namespace Drawbridge.Core;

public sealed record HttpRequestData(string Method, string Url, Dictionary<string, string> Headers, string? Body);

public sealed record HttpResponseData(int Status, string Body);

/// <summary>Thrown by an IHttpClient when the request times out.</summary>
public sealed class RequestTimeoutException() : Exception("request timed out");

public interface IHttpClient
{
    Task<HttpResponseData> SendAsync(HttpRequestData req, int timeoutMs);
}

/// <summary>Default client backed by System.Net.Http. Maps a timeout to RequestTimeoutException.</summary>
public sealed class FetchClient(HttpClient? http = null) : IHttpClient
{
    private readonly HttpClient _http = http ?? new HttpClient();

    public async Task<HttpResponseData> SendAsync(HttpRequestData req, int timeoutMs)
    {
        using var cts = new CancellationTokenSource(timeoutMs);
        using var msg = new HttpRequestMessage(new HttpMethod(req.Method), req.Url);
        foreach (var (k, v) in req.Headers)
        {
            if (k.Equals("content-type", StringComparison.OrdinalIgnoreCase)) continue;
            msg.Headers.TryAddWithoutValidation(k, v);
        }
        if (req.Body is not null)
        {
            // Bare "application/json" (no "; charset=utf-8") to match Node's fetch header.
            msg.Content = new StringContent(req.Body, System.Text.Encoding.UTF8);
            msg.Content.Headers.ContentType = new System.Net.Http.Headers.MediaTypeHeaderValue("application/json");
        }

        try
        {
            using var res = await _http.SendAsync(msg, cts.Token);
            return new HttpResponseData((int)res.StatusCode, await res.Content.ReadAsStringAsync(cts.Token));
        }
        catch (OperationCanceledException) when (cts.IsCancellationRequested)
        {
            throw new RequestTimeoutException();
        }
    }
}

File: src/node/src/exec/http.ts
export interface HttpRequest {
  method: string;
  url: string;
  headers: Record<string, string>;
  body?: string;
}

export interface HttpResponse {
  status: number;
  body: string;
}

export class TimeoutError extends Error {
  constructor() {
    super("request timed out");
    this.name = "TimeoutError";
  }
}

/** Sends a request; throws TimeoutError on timeout. Injected so tests use a stub. */
export type HttpClient = (req: HttpRequest, timeoutMs: number) => Promise<HttpResponse>;

/** Default client backed by global fetch + AbortController. */
export const fetchClient: HttpClient = async (req, timeoutMs) => {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const res = await fetch(req.url, {
      method: req.method,
      headers: req.headers,
      body: req.body,
      signal: controller.signal,
    });
    return { status: res.status, body: await res.text() };
  } catch (e) {
    if (e instanceof Error && e.name === "AbortError") throw new TimeoutError();
    throw e;
  } finally {
    clearTimeout(timer);
  }
};
